src/ray/core_worker/task_interface.cc

```diff
@@ -100,64 +98,63 @@ CoreWorkerTaskInterface::CoreWorkerTaskInterface(
                                new CoreWorkerRayletTaskSubmitter(raylet_client)));
 }
 
-Status CoreWorkerTaskInterface::SubmitTask(const RayFunction &function,
-                                           const std::vector<TaskArg> &args,
-                                           const TaskOptions &task_options,
-                                           std::vector<ObjectID> *return_ids) {
-  auto &context = worker_context_;
-  auto next_task_index = context.GetNextTaskIndex();
-  const auto task_id = GenerateTaskId(context.GetCurrentJobID(),
-                                      context.GetCurrentTaskID(), next_task_index);
+raylet::TaskSpecBuilder CoreWorkerTaskInterface::BuildCommonTaskSpec(
+    const RayFunction &function, const std::vector<TaskArg> &args, uint64_t num_returns,
+    const std::unordered_map<std::string, double> &required_resources,
+    const std::unordered_map<std::string, double> &required_placement_resources,
+    std::vector<ObjectID> *return_ids) {
+  raylet::TaskSpecBuilder builder;
+  auto next_task_index = worker_context_.GetNextTaskIndex();
+  // Build common task spec.
+  builder.SetCommonTaskSpec(
+      function.language, function.function_descriptor, worker_context_.GetCurrentJobID(),
+      worker_context_.GetCurrentTaskID(), next_task_index, num_returns,
+      required_resources, required_placement_resources);
+  // Set task arguments.
+  for (const auto &arg : args) {
+    if (arg.IsPassedByReference()) {
+      builder.AddByRefArg(arg.GetReference());
+    } else {
+      builder.AddByValueArg(arg.GetValue()->Data(), arg.GetValue()->Size());
+    }
+  }
 
-  auto num_returns = task_options.num_returns;
+  // Compute return IDs.
+  const auto task_id = TaskID::FromBinary(builder.GetMessage().task_id());
   (*return_ids).resize(num_returns);
```

Comment:
better to return_ids-> resize()

Reply:
why is that better?

Reply:
just have a better readability I think. 
for readers, `return_ids->resize()` is only one operator.